### PR TITLE
refactor: update the volume create api

### PR DESCRIPTION
**Pull request by tiagolobocastro**
This follows the new volume api described in the design docs.
Topology information may now be filled in via the operator, either as an explicit
selection of nodes, or using labels for nodes and pools.

Before the labels can be used the node and pool will have to be updated to include
them at creation time and also via an update method.

### CODE REVIEW

File: control-plane/mbus-api/src/v0.rs [hunk 1]
```diff
@@ -1161,22 +1246,26 @@ pub struct CreateVolume {
...
-    #[serde(default)]
-    pub allowed_nodes: Vec<NodeId>,
-    /// preferred nodes for the replicas
-    #[serde(default)]
-    pub preferred_nodes: Vec<NodeId>,
-    /// preferred nodes for the nexuses
-    #[serde(default)]
-    pub preferred_nexus_nodes: Vec<NodeId>,
+    /// replica replacement policy
+    pub policy: VolumeHealPolicy,
```
Comment (paulyoong) [inline control-plane/mbus-api/src/v0.rs:1252]: The comment mentions replica but the type is volume which is a little confusing. I think the comment should say volume. The user can set the policy on the volume, but internally we manage it by manipulating the replicas.

**Review by paulyoong — APPROVED** [review]
Looks good.

From what I can see we are dropping the ability for the user to set the number of nexuses for now. Is that correct?

File: control-plane/mbus-api/src/v0.rs [hunk 2]
```diff
@@ -1143,6 +1143,91 @@ pub struct Volume {
...
+/// Volume topology used to determine how to place/distribute the data
+/// Should either be labelled or explicit, not both.
+/// If neither is used then the control plane will select from all available resources.
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq, Apiv2Schema)]
+pub struct Topology {
+    /// volume topology using labels
+    pub labelled: Option<LabelTopology>,
+    /// volume topology, explicitly selected
+    pub explicit: Option<ExplicitTopology>,
+}
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/mbus-api/src/v0.rs:1164]: No, sadly enum variants don't work for openapiv2 and paperclip does not support them.

File: control-plane/mbus-api/src/v0.rs [hunk 3]
```diff
@@ -1143,6 +1143,91 @@ pub struct Volume {
...
+/// Excludes resources with the same $label name, eg:
+/// "Zone" would not allow for resources with the same "Zone" value
+/// to be used for a certain operation, eg:
+/// A node with "Zone: A" would not be paired up with a node with "Zone: A",
+/// but it could be paired up with a node with "Zone: B"
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq, Apiv2Schema)]
+pub struct ExclusiveLabel(
+    /// exclusive label NAME in the form "NAME", and not "NAME: VALUE"
+    pub String,
+);
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/mbus-api/src/v0.rs:1175]: Good point

File: control-plane/mbus-api/src/v0.rs [hunk 4]
```diff
@@ -1143,6 +1143,91 @@ pub struct Volume {
...
+pub struct PoolTopology {
+    /// inclusive labels
+    #[serde(default)]
+    pub inclusion: Vec<InclusiveLabel>,
+}
+
+/// Explicit node placement Selection for a volume
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq, Apiv2Schema)]
+pub struct ExplicitTopology {
+    /// replicas can only be placed in these nodes
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/mbus-api/src/v0.rs:1212]: Done

File: control-plane/mbus-api/src/v0.rs [hunk 5]
```diff
@@ -1143,6 +1143,91 @@ pub struct Volume {
...
+    #[serde(default)]
+    pub allowed_nodes: Vec<NodeId>,
+    /// preferred nodes to place the replicas
+    #[serde(default)]
+    pub preferred_nodes: Vec<NodeId>,
+}
+
+/// Replica policy used by the volume to determine if and how to replace a replica
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/mbus-api/src/v0.rs:1220]: Yep

File: control-plane/mbus-api/src/v0.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/mbus-api/src/v0.rs:1252]: Yep I had it all as replica then replaced with volume but clearly missed some :)

File: control-plane/mbus-api/src/v0.rs [hunk 6]
```diff
@@ -1143,6 +1143,91 @@ pub struct Volume {
...
+    #[serde(default)]
+    pub inclusion: Vec<InclusiveLabel>,
+}
+
+/// Placement pool topology used by volume operations
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq, Apiv2Schema)]
+pub struct PoolTopology {
+    /// inclusive labels
+    #[serde(default)]
+    pub inclusion: Vec<InclusiveLabel>,
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply control-plane/mbus-api/src/v0.rs:1206]: I'm not sure, I'll have to review the wiki pages to see if there's any info about this there

**Comment by tiagolobocastro** [comment]
> Looks good.
> 
> From what I can see we are dropping the ability for the user to set the number of nexuses for now. Is that correct?

Since we're not using yet I left it out for the moment to keep things simple, but I can add it back if you think we should have it?

**Comment by paulyoong** [comment]
> > Looks good.
> > From what I can see we are dropping the ability for the user to set the number of nexuses for now. Is that correct?
> 
> Since we're not using yet I left it out for the moment to keep things simple, but I can add it back if you think we should have it?

No need - I just wanted to confirm that was intended.

**Comment by tiagolobocastro** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
Changing the types definitions also changes the openapi spec so we might want to add those paths to the commit linter...
bors merge

**Pull request merged**